test/integration/sequelize/deferrable.test.js

```diff
@@ -0,0 +1,126 @@
+'use strict';
+
+/* jshint -W030 */
+/* jshint -W110 */
+var _ = require('lodash')
+  , chai = require('chai')
+  , expect = chai.expect
+  , assert = chai.assert
+  , Support = require(__dirname + '/../support')
+  , dialect = Support.getTestDialect()
+  , Sequelize = require(__dirname + '/../../../index')
+  , config = require(__dirname + '/../../config/config')
+  ;
+
+describe(Support.getTestDialectTeaser('Sequelize'), function() {
+  if (dialect !== 'postgres') {
+    return;
+  }
+
+  beforeEach(function () {
+    this.run = function (deferrable, options) {
+      options = options || {};
+
+      var taskTableName      = options.taskTableName || 'tasks_' + config.rand();
+      var transactionOptions = _.extend({ deferrable: Sequelize.Deferrable.SET_DEFERRED }, options);
+      var userTableName      = 'users_' + config.rand();
+
+      var User = this.sequelize.define(
+        'User', { name: Sequelize.STRING }, { tableName: userTableName }
+      );
+
+      var Task = this.sequelize.define(
+        'Task', {
+          title: Sequelize.STRING,
+          user_id: {
+            allowNull: false,
+            type: Sequelize.INTEGER,
+            references: userTableName,
+            referencesKey: 'id',
+            referencesDeferrable: deferrable
+          }
+        }, {
+          tableName: taskTableName
+        }
+      );
+
+      return User.sync({ force: true }).bind(this).then(function () {
+        return Task.sync({ force: true });
+      }).then(function () {
+        return this.sequelize.transaction(
+          transactionOptions,
+          function (t) {
+            return Task
+              .create({ title: 'a task', user_id: -1 }, { transaction: t })
+              .then(function (task) {
+                return [task, User.create({}, { transaction: t })];
+              })
+              .spread(function (task, user) {
+                task.user_id = user.id;
+                return task.save({ transaction: t });
+              });
+          }
+        );
+      });
+    };
+  });
+
+  describe('Deferrable', function () {
+    describe('NOT', function () {
+      it('does not allow the violation of the foreign key constraint', function () {
+        return this
+          .run(Sequelize.Deferrable.NOT)
+          .then(assert.fail, function (err) {
```

Comment:
It's about the [Promise anti-pattern] (https://github.com/petkaantonov/bluebird/wiki/Promise-anti-patterns#the-thensuccess-fail-anti-pattern).

But what Mick said is even better here, so it would be good to write it as:
```js
return expect(this.run(Sequelize.Deferrable.NOT)).to.eventually.be.rejectedWith(Sequelize.ForeignKeyConstraintError);
```